gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -301,20 +316,24 @@ private WriteObjectResponse doResumableUpload() throws IOException {
                   ? responseObserver.transientError
                   : responseObserver.nonTransientError);
           break;
-        } else if (objectFinalized) {
+        }
+        if (objectFinalized
+            || requestChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
+          // close request stream if final request OR Map is full
           requestStreamObserver.onCompleted();
+          try {
+            responseObserver.done.await();
+            requestChunkMap.clear();
```

Comment:
With this we are losing track of the offsets which are yet to be committed. What are the implications.

Reply:
No, we are not we are cleaning up iff we have got a successful response from server (via ResponseObserver).